### PR TITLE
Initial PromiseKit Integration

### PR DESCRIPTION
**Pull request by keefertaylor**
Begin utilizing promises in TezosNodeClient. Future CLs will compose additional promises to remove all callback handlers.

### CODE REVIEW

**Comment by keefertaylor** [comment]
The diff on github is pretty gnarly looking but is actually really simple. 

Internals of `getMetadataForOperation` gets wrapped in a new promise which is fulfilled or rejected based on the result. `forgeSignPreapplyAndInjectOperations` waits on the promise and wraps the rest of the code.

File: TezosKit.podspec [hunk 1]
```diff
@@ -17,7 +17,8 @@ Pod::Spec.new do |s|
 
   s.dependency "BigInt", "~> 3.1"		
   s.dependency "MnemonicKit"
-  s.dependency "Sodium", "~> 0.7.0"  
+  s.dependency "PromiseKit"
```
Comment (jlaws) [inline TezosKit.podspec]: I would reconsider adding any dependencies to the library, especially one like PromiseKit.  The common pattern is to make subspecs that update your public API's to be friendly with other libraries.  In this case you could have a TezosKit/PromiseKit subspec that adds additional promisekit friendly API's to the standard ones. You could also create a TezosKit/Reactive to make an RXSwift friendly API.

File: TezosKit/Client/TezosNodeClient.swift [hunk 2]
```diff
@@ -538,70 +537,73 @@ public class TezosNodeClient: AbstractClient {
...
-      if let json = json,
-        let fetchedChainID = json["chain_id"] as? String,
-        let fetchedHeadHash = json["hash"] as? String,
-        let fetchedProtocolHash = json["protocol"] as? String {
-        chainID = fetchedChainID
-        headHash = fetchedHeadHash
-        protocolHash = fetchedProtocolHash
+  private func getMetadataForOperation(address: String) -> Promise<OperationMetadata> {
+    return Promise { seal in
+      let fetchersGroup = DispatchGroup()
```
Comment (jlaws) [inline TezosKit/Client/TezosNodeClient.swift]: don't use dispatch groups anymore when using promises, you should update GetChainHeadRPC to return a promise and use a 
when(fulfilled: ) 
to combine all the promises

File: TezosKit/Client/TezosNodeClient.swift [hunk 3]
```diff
@@ -538,70 +537,73 @@ public class TezosNodeClient: AbstractClient {
...
+      fetchersGroup.enter()
+      send(rpc: chainHeadRequestRPC)
+
+      fetchersGroup.enter()
+      send(rpc: getAddressCounterRPC)
+
+      fetchersGroup.enter()
+      send(rpc: getAddressManagerKeyRPC)
+
+      fetchersGroup.wait()
```
Comment (jlaws) [inline TezosKit/Client/TezosNodeClient.swift]: Update this to use all promises and not need any dispatch groups

Comment (keefertaylor) [reply TezosKit/Client/TezosNodeClient.swift]: Acknowledged.

File: TezosKit/Client/TezosNodeClient.swift (hunk 2 above)
Comment (keefertaylor) [reply TezosKit/Client/TezosNodeClient.swift]: Done. I can't update this code if I'm keeping promises outside of the core library, so left as is. 

I do realize what you are talking about though, I was going to try to stage promisekit integration out over a few CLs in order to make it easier to review.

**Comment by keefertaylor** [comment]
PTAL. 

Apologies that this has ended up touching so many files. The main parts are:
- Decouple completion blocks from RPC objects and instead pass them through TezosNodeClient.
- Add Extensions/PromiseKit/*
- Rebase Tests/* to Tests/TezosKit/* and add Tests/Extensions/PromiseKit/*
- Wrap the completion block API with a promise in the PromiseKit extension
- Add unit tests in the API for new PromiseKit functionality

TezosNodeClient should also receive a PromiseKit extension. I'll do that in a follow up PR to keep this one vaguely manageable.

**Pull request merged**